### PR TITLE
路線判定: 駅部の線形欠落で並走路線へ誤ロックする問題を修正

### LINKED ISSUE #67: fix: route matching locks parallel Shinkansen / subway / private lines on JR corridors (Tabata, Tokyo-Ueno, Shinagawa)
作成日: 2026-09-15

関連PR: #66（この不具合を検出した回帰テスト）

対象コード:

- `src/domain/railway/map-matcher.ts` — 初回ロック・挑戦者判定・再取得（`considerInitialLock` / `updateChallenger` / `considerReacquireLock`）
- `src/domain/railway/candidate-scorer.ts` — 距離・方位・連続性・速度ペナルティ
- `src/domain/railway/window-scorer.ts` / `route-health.ts`
- `src/config/tracking-config.ts`
- データ: 同梱サンプル `src/data/sample/track-segments.json` の `seg-shinkansen-tokyo-ueno`（3点の直線近似）

## 1. 事象概要

PR #66 の回帰テスト（`pnpm test:route-edge`、実データ v1.4.0 の線形＋合成GPS、`DEFAULT_TRACKING_CONFIG`）で、JRの在来線に乗っているのに、並走する別路線がHUDに表示される事象を確認した。`KNOWN_ISSUES` に 13 シナリオ分を記録している。

| 区間 | 実際の乗車 | 表示された路線 | 程度 |
| --- | --- | --- | --- |
| 田端（北行き・合流側） | 京浜東北線 / 山手線 | 東北新幹線（大宮〜東京） | 西日暮里〜田端から切り替わり、京浜東北線北行きと合流側では最後まで戻らない（誤表示 140〜330秒） |
| 東京〜上野 | 京浜東北線 | JR東北新幹線（同梱サンプルの直線近似） | ほぼ全区間（誤表示 210〜370秒） |
| 品川付近 | 山手線 / 京浜東北線 | 京急「本線」 | 46〜90秒 |
| 東京〜新橋 | 京浜東北線 | 8号線有楽町線 | 約55秒 |
| 御茶ノ水 | 中央・総武線各駅停車 | 4号線丸ノ内線 | 約58秒 |
| 両国 | 中央・総武線各駅停車 | 総武線（両国〜東京）＝快速線 | 17〜48秒 |

いずれも走行速度は在来線の範囲（〜90km/h）で、新幹線を表示する合理性はない。

## 2. 再現手順

```bash
pnpm test:route-edge
```

`tests/railway/route-edge-cases/route-edge-cases.test.ts` の `KNOWN_ISSUES`（`[known issue]` 付きのテスト）と、`__snapshots__` の各シナリオの `runs` 内訳で秒数を確認できる。

## 3. 原因の仮説（未検証）

- **東京〜上野**: 同梱サンプル `seg-shinkansen-tokyo-ueno` は東京→上野を3点の直線で結んでおり、秋葉原付近で在来線の線路帯上を通る。MLIT の東北新幹線（大宮〜東京）と二重に登録され、本番でも両方がマージされる。
- **田端・東京〜上野**: 候補スコアが主に「距離・方位・連続性」で決まり、線路間隔が GPS 誤差以下の並走区間では新幹線・地下鉄・私鉄と区別できない。速度ペナルティは 160km/h 超でしか働かない。
- **ロック後に戻らない**: 一度誤路線にロックすると、連続性ボーナスと挑戦者判定のマージン条件により、正しい路線が挑戦者として優勢にならない可能性。
- **地下鉄・私鉄**: 駅停車中（速度≈0・方位なし）に距離だけで比較され、近接する別事業者の路線へ移る可能性。

## 4. 期待する挙動

- 在来線の速度・停車パターンで走行中は、並走する新幹線を表示しない。
- 路線帯が明確に分かれた後は、正しい路線に復帰する。
- 確信が弱い場合は、誤った路線を表示するより未確定（非表示）を優先する（`docs/ROUTE_MATCHING_RELIABILITY_SPEC.md` の方針）。

## 5. 受け入れ条件

- `KNOWN_ISSUES` の項目が減る（改善した項目は削除し、スナップショットを更新）。
- `KNOWN_ISSUES` にないハードチェックが新たに失敗しない。
- 新幹線乗車シナリオ（`tokyo-shinagawa-tokaido-shinkansen`、逆方向含む）で引き続き新幹線を判定できる。
- 既存テスト（`tests/railway/asakusabashi-sobu.test.ts` ほか）が通る。
- 改善していないシナリオのスナップショット指標（誤表示秒数・路線変更回数・初回表示までの秒数）が悪化しない。悪化がある場合はトレードオフとしてPRに明記する。

### PR DESCRIPTION
**Pull request by TakuroFukamizu**
Fixes #67

> #66（回帰テスト）を土台にした stacked PR です。#66 のマージ後に base を `main` へ変更してください。

## 背景

#66 の回帰テストで、JR在来線に乗車中に並走する東北新幹線・同梱の直線近似「JR東北新幹線」・京急本線・地下鉄・総武快速線が表示される事象を確認した（#67）。

## 原因

4つの誤判定クラスタを並列に診断した結果、共通の引き金は**駅部での線形欠落**だった。

- MLIT の在来線セグメントはホームではなく分岐点で途切れるため、駅に 200〜520m の穴がある（田端 519m、西日暮里 203m、品川 422m、有楽町 380m、御茶ノ水 368m、両国 437m など）。
- 停車中や穴の通過中は、ロック中セグメントまでの距離が端点へのクランプで 100〜370m に膨らむ。route health が 0.45 を下回り、SUSPICIOUS → REACQUIRING に遷移する。
- その間、連続した線形を持つ別路線（新幹線・地下鉄・私鉄・快速線）が唯一の近い候補になり、ロックを奪う。
- 一度奪うと、同一セグメントの連続性ボーナス +20 が距離差（30m 並走で約 12 点）を上回り、正しい路線に戻れない。

## 変更

**スコアリング（主修正）**
- `src/domain/railway/segment-gap.ts`（新規）: ロック中・隣接セグメントの端点を越えた地点を、端点接線の延長、または位置の揃った対岸セグメントへの橋渡しで評価する（600m・45° 以内）。継続セグメントや他路線の連続線形がフィックスを明確に覆う場合は適用しない。
- `candidate-scorer.ts`: 上記を前回セグメントと隣接セグメントにのみ適用。
- `window-scorer.ts`: 窓内の距離減衰を GPS 精度で正規化。ただし従来の 120m 線形減衰より緩くはしない（低精度時のタイブレークを維持）。

**ロック状態の正しさ（別案から移植、単独で計測し悪化なしのものだけ採用）**
- 切替直後のルートは、デフォルト値で組まれた health による一発ロックをせず、確認（連続フィックス）を先に行う。
- OS が停車を報告している間は、挑戦者の勝ち数と経過時間を凍結する（停車は路線の証拠にならない）。GPS 喪失時は凍結しない。
- 新しい閾値はすべて `TrackingConfig` に移した（仕様 §7「閾値をハードコードしない」）。

**不採用**
- window 支持切替へのマージン必須化: 東京〜上野で JR東北新幹線が再び表示される（wrongLineS 26→459）ため取り下げた。

## 結果（回帰テスト 20 シナリオ合計）

| 指標 | 修正前 | 修正後 |
| --- | --- | --- |
| 誤表示秒数 `wrongLineS` | 2006 | 23 |
| 禁止路線（並走新幹線）表示秒数 `forbiddenLineS` | 1456 | 0 |
| 初回表示までの秒数 `firstDisplayAtS` | 654 | 654（変化なし） |
| 路線変更回数 `lineChanges` | 40 | 28 |

- `KNOWN_ISSUES`: 13 シナリオ → 3 シナリオ。残りはすべて `no-wrong-line` で、分岐後に直前の正しい路線を 45 秒の判定猶予を少し超えて保持するもの。
- 東海道新幹線の乗車シナリオ（往復）は変化なし。引き続き新幹線を判定する。
- 低精度 GPS（トンネル相当 σ30m）での誤表示秒数も 1696 → 1425 に改善し、悪化していない。

## 検証の進め方（ultracode）

1. 4 クラスタを並列診断し、tick 単位で根本原因を実測。
2. スコアリング／ロック状態／データ・トポロジーの 3 案を別 worktree で実装・計測。3 名の審査で本 PR のスコアリング案を採用した（他案は新幹線乗車の復帰遅延や、合成トレースに合わせた過適合のリスクあり）。
3. 別案の一般的な修正を 1 つずつ移植して計測。
4. 正しさ・本番リスク・テスト完全性・仕様整合の 4 観点でレビューし、指摘を個別に反証検証。確認された 6 件のうち 5 件を修正した。
   - 停車時間が挑戦者の継続時間に算入される問題
   - 低精度時にタイブレークできない問題
   - 他路線の連続線形があっても穴越しにロックを守り続ける問題（部分修正）
   - テストの識別力不足
   - 閾値のハードコード

## テスト

```
CI=true pnpm test     Test Files 44 passed / Tests 418 passed
pnpm lint / npx tsc --noEmit: エラーなし
```

- 追加テスト: `tests/railway/segment-gap.test.ts`、`route-switch-guards.test.ts`、`window-scorer-degraded-gps.test.ts`。各テストはガードを外すと失敗することを確認済み（ミューテーション確認）。
- 回帰テストの許容路線・判定猶予・トレース・フィクスチャ、および既存テスト（`asakusabashi-sobu` / `map-matcher` / `route-matching-review`）は変更していない。

## 残課題・リスク

- **合成データでしか検証していない**: 実機 GPS（マルチパスによる横ずれ、疎なフィックス）での確認が必要。穴越し投影は、誤ってロックした路線も最大 600m 守りうる。
- **品川での京急への乗り換え**: 品川から約 400m は JR の接線投影と京急が 1 精度以内で拮抗し、切替が遅れる（部分修正）。別の仕組みでの対応が必要。
- **データ側の問題（未対応）**: 同梱サンプル `seg-shinkansen-tokyo-ueno` の二重登録と、東海道新幹線 東京〜品川の欠落。

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_0141t9HP7dCpCJ1Mx1qiwz3b

### CODE REVIEW

File: src/domain/railway/map-matcher.ts [hunk 1]
```diff
@@ -605,9 +608,27 @@ export class MapMatcher {
...
+    // While the OS says the train is stopped (a dwell, often inside a station hole of
+    // the locked segment) a fix says nothing about which line it is on: it neither adds
+    // challenger wins nor resets them. An unreliable trajectory (GPS loss) does not freeze.
+    // The challenger's clock is paused too: time spent stopped is not persistence
+    // evidence, so it must not count toward routeChallengerMinimumMs.
+    if (osStopped === true) {
+      if (this.challenger) {
+        const pausedMs = Math.max(0, nowMs - this.challenger.lastSeenAtMs);
+        this.challenger.firstSeenAtMs += pausedMs;
+        this.challenger.lastSeenAtMs += pausedMs;
```
[the remark this answers is not in the document]
Comment (TakuroFukamizu) [reply src/domain/railway/map-matcher.ts]: 修正しました（5293163）。`firstSeenAtMs` / `lastSeenAtMs` は実時刻のまま維持し、停車を除いた経過時間を `activeDurationMs` として別に持たせ、`routeChallengerMinimumMs` の判定にはそちらを使うようにしました。デバッグパネルは「走行中」と「実時間」の両方を表示します。

File: src/domain/railway/candidate-scorer.ts [hunk 2]
```diff
@@ -32,11 +33,29 @@ export function scoreCandidate(input: ScoreCandidateInput): RouteCandidateScore
...
   const accuracyFloor = Math.max(sample.accuracyMeters, config.routeMinimumAccuracyMeters);
+
+  // Station holes: segments end at junctions, not platforms. For the current segment (or
+  // one attached to it) measure the offset from where the track continues through the hole
+  // instead of the growing straight-line gap to the end vertex.
+  if (previousSegment && (previousSegment.id === segment.id || segmentsAreAdjacent(previousSegment, segment))) {
+    const gap = projectAcrossEndGap(sample, segment, closest, nearbySegments, {
+      maxOverrunMeters: config.routeSegmentEndOverrunMeters,
+      coverToleranceMeters: accuracyFloor * config.routeSegmentGapCoverAccuracyMultiple,
+      otherCoverMarginMeters: accuracyFloor * config.routeSegmentGapOtherCoverMarginAccuracyMultiple,
```
[the remark this answers is not in the document]
Comment (TakuroFukamizu) [reply src/domain/railway/candidate-scorer.ts]: 計測した結果、今回は変更しないことにしました。

回帰テスト20シナリオ（10,440フィックス、`findSegmentsNear` が返す近傍セグメントは最大61本）を `routeSegmentEndOverrunMeters` の有無で比較しました。

- 投影あり（600m）: 1フィックスあたり 0.308ms
- 投影なし（0m）: 1フィックスあたり 0.343ms

差は計測ノイズの範囲で、投影による増加は確認できませんでした。`projectAcrossEndGap()` の呼び出しは前回セグメントと隣接セグメントに限定されており、内部の走査も端点付近で早期に打ち切られるためと考えられます。GPS は 1Hz なので、この規模では実機でも支障はないと判断しています。

実機での確認（#67 の修正を含むビルドで実施予定）でプロファイル上の問題が出た場合は、フィックスごとの最近傍点キャッシュ導入を検討します。
